content/blog/end-to-end-encryption/index.md

```diff
@@ -0,0 +1,125 @@
+
+You could save the file, send it to the other person, have them open it. But that's pretty cumbersome. In this post we'll see how to get a link working without the server having access to the data.
+
+## Traditional Website Architecture
+
+In the traditional website architecture, if you want to save a scene, you're going to send the content to the server so that it can store it. Then to load it you'll the server to send it back to you.
+
+In that world, you trust the server to contain your information but you don't trust the pipes in-between the client and server using https to encrypt the data.
+
+[![Untrusted Network](untrusted-network.png)](https://excalidraw.com/#json=5649116445016064,yOfExolZoMhtGnysT3-LWA)
+
+This works well unless the server gets compromised. The attacker will have access to every single drawing ever made! This is something we'd like to avoid.
+
+## End-to-End Encryption
+
+WhatsApp popularized end-to-end encryption, a technique that allows various clients to communicate without the server being able to read the content of the communication.
+
+The idea is to encrypt the content before sending it to the server. The server would just store the encrypted blob and send it back to the client.
+
+[![Untrusted Server](untrusted-server.png)](https://excalidraw.com/#json=5645858175451136,8w-G0ZXiOfRYAn7VWpANxw)
+
+The biggest challenge with this architecture is how to distribute the key to encrypt the message in such a way that the server cannot see it.
+
+Thankfully, in the context of a website, we can exploit the hash part of the URL. Anything that's added after the `#` doesn't get sent to the server, but is readable from the client-side JavaScript code.
+
+[![url](url.png)](https://excalidraw.com/#json=5660568841093120,vki3y9xuEulFVHDqt-PBMw)
+
+
+## Show me the code
+
+Fortunately, the Web Cryptography APIs are now widely available to all the browsers that let us implement this. That said, the APIs to deal with encryption, keys and binary data are not the most straightforward, this next section walks you through how to wire it all together.
+
+### Upload
+
+We generate a random key that will be used to encrypt the data.
+
+```javascript
+const key = await window.crypto.subtle.generateKey(
+  { name: "AES-GCM", length: 128 },
+  true, // extractable
+  ["encrypt", "decrypt"],
+);
+```
+
+We encrypt the content with that random key. In this case, we only encrypt the content once with the random key so we don't need an `iv` and can leave it filled with 0.
+
+```javascript
+const encrypted = await window.crypto.subtle.encrypt(
+  { name: "AES-GCM", iv: new Uint8Array(12) /* don't reuse key! */ },
+  key,
+  new TextEncoder().encode(JSON.stringify(content)),
+);
+```
+
+We upload the encrypted content to the server. Note that we don't send the key to the server!
+
+```javascript
+const response = await (await fetch("upload.php", {
```

Comment:
can we get rid of PHP? I'm not too religious about it, but we are not using PHP anywhere in Excalidraw.

Reply:
I tried to show that it's an endpoint. I guess I'm old school... I'll remove